### PR TITLE
[performance] reduce dependency resolution git calls

### PR DESCRIPTION
**Pull request by tomerd**
motivation: better performance, happier users

changes:
* the bottle neck in dependency resolution is the iteration of available versions and reducing the list to thos with matching tools-versions,
  each version iteration shells out to git several times now, which is fairly expensive given the shelling out aspect.
* change RepositoryPackageContainer::toolsVersion to find & load the manifest based on tag
* change RepositoryPackageContainer::loadManifest to load the manifest based on tag
* this reduces the amount of git shelling (in this use case by 60%), reducing total resolutoin duration  for large packages by 25%

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please test package compatibility

File: Sources/PackageGraph/RepositoryPackageContainer.swift [hunk 1]
```diff
@@ -146,9 +146,8 @@ public class RepositoryPackageContainer: PackageContainer, CustomStringConvertib
             guard let tag = try self.knownVersions()[version] else {
                 throw StringError("unknown tag \(version)")
             }
-            let revision = try repository.resolveRevision(tag: tag)
```
Comment (tomerd) [inline Sources/PackageGraph/RepositoryPackageContainer.swift:149]: this cuts 2 shell calls

File: Sources/PackageGraph/RepositoryPackageContainer.swift [hunk 2]
```diff
@@ -158,8 +157,7 @@ public class RepositoryPackageContainer: PackageContainer, CustomStringConvertib
                 guard let tag = try self.knownVersions()[version] else {
                     throw StringError("unknown tag \(version)")
                 }
-                let revision = try repository.resolveRevision(tag: tag)
```
Comment (tomerd) [inline Sources/PackageGraph/RepositoryPackageContainer.swift:161]: this cuts 2 shell calls

**Review by neonichu — APPROVED** [review]

**Pull request merged**